Mathlib/AlgebraicTopology/SimplicialSet/StrictSegal.lean

```diff
@@ -0,0 +1,156 @@
+-/
+
+universe v u
+
+open CategoryTheory
+
+open Simplicial SimplexCategory
+
+namespace SSet
+
+variable (X : SSet.{u})
+
+/-- A simplicial set `X` satisfies the strict Segal condition if its simplices are uniquely
+determined by their spine. -/
+class StrictSegal : Prop where
+  segal : ∀ (n : ℕ), Function.Bijective (X.spine n)
+
+variable {X} in
+/-- The diagonal of a simplex is the long edge of the simplex.-/
+def diagonal {n : ℕ} (Δ : X _[n]) : X _[1] := X.map ((mkOfDiag n).op) Δ
+
+namespace StrictSegal
+variable {X : SSet.{u}} [StrictSegal X] {n : ℕ}
+
+/-- In the presence of the strict Segal condition, a path of length `n` extends to an `n`-simplex
+whose spine is that path. -/
+noncomputable def spineToSimplex : Path X n → X _[n] :=
+  (Equiv.ofBijective _ (segal n)).invFun
+
+@[simp]
+theorem spineToSimplex_spine (f : Path X n) :
+    X.spine n (StrictSegal.spineToSimplex f) = f :=
+  (Equiv.ofBijective _ (segal n)).right_inv f
+
+@[simp]
+theorem spineToSimplex_vertex (i : Fin (n + 1)) (f : Path X n) :
+    X.map (const [0] [n] i).op (spineToSimplex f) = f.vertex i := by
+  rw [← spine_vertex, spineToSimplex_spine]
+
+@[simp]
+theorem spineToSimplex_spine_edge (i : Fin n) (f : Path X n) :
+    X.map (mkOfSucc i).op (spineToSimplex f) = f.arrow i := by
+  rw [← spine_arrow, spineToSimplex_spine]
+
+
+/-- In the presence of the strict Segal condition, a path of length `n` can be "composed" by taking
+the diagonal edge of the resulting `n`-simplex. -/
+noncomputable def spineToDiagonal (f : Path X n) : X _[1] := diagonal (spineToSimplex f)
+
+@[simp]
+theorem spineToSimplex_interval (f : Path X n) (j l : ℕ) (hjl : j + l < n + 1)  :
+    X.map (subinterval j l hjl).op (spineToSimplex f) =
+      spineToSimplex (Path.interval f j l hjl) := by
+  apply (segal _).injective
+  rw [StrictSegal.spineToSimplex_spine]
+  ext i
+  · simp only [mkHom, Equiv.invFun_as_coe, spine_vertex, Fin.coe_addNat,
+      ← FunctorToTypes.map_comp_apply, ← op_comp, const_comp, len_mk, spineToSimplex_vertex]
+    congr
+    exact Fin.eq_of_val_eq (Nat.mod_eq_of_lt (by omega)).symm
+  · unfold Path.interval
+    simp only [Equiv.invFun_as_coe, spine_arrow, Fin.coe_addNat]
+    simp only [← FunctorToTypes.map_comp_apply, ← op_comp]
+    have ceq : mkOfSucc i ≫ subinterval j l hjl = mkOfSucc ⟨i + j, (by omega)⟩ := by
+      ext ⟨e, he⟩ : 3
+      unfold subinterval
+      match e with
+      | 0 => rfl
+      | 1 => ?_
+      conv_rhs =>
+        apply mkOfSucc_homToOrderHom_one
+      simp only [len_mk, Nat.reduceAdd, mkHom, comp_toOrderHom, Hom.toOrderHom_mk, Fin.mk_one,
+        Fin.isValue, OrderHom.comp_coe, OrderHom.coe_mk, Function.comp_apply, Fin.succ_mk,
+        Fin.mk.injEq]
+      exact Nat.succ_add_eq_add_succ ↑i ↑j
+    rw [ceq]
+    simp only [spineToSimplex_spine_edge]
+
+@[simp]
+theorem spineToSimplex_edge (f : Path X n) (j l : ℕ) (hn : j + l < n + 1) :
+    X.map (mkOfLe ⟨j, (by omega)⟩ ⟨j + l, hn⟩ (Nat.le_add_right j l)).op (spineToSimplex f) =
```

Comment:
This hints that we might want to abstract `mkOfLe ⟨j, (by omega)⟩ ⟨j + l, hn⟩ (Nat.le_add_right j l)` into a definition `edge j l hn`. What do you think?

Reply:
Done. I've called this `intervalEdge` but please suggest a different name.